client/extensions/woocommerce/woocommerce-services/components/shipping-classes-field/index.js

```diff
@@ -0,0 +1,93 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import React from 'react';
+import PropTypes from 'prop-types';
+import { uniqBy } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import FormFieldset from 'components/forms/form-fieldset';
+import FormLabel from 'components/forms/form-label';
+import TokenField from 'components/token-field';
+import FieldError from '../field-error';
+import FormSettingExplanation from 'components/forms/form-setting-explanation';
+
+export default class ShippingClassesField extends React.Component {
+	static propTypes = {
+		id: PropTypes.string.isRequired,
+		title: PropTypes.string,
+		description: PropTypes.string,
+		value: PropTypes.array.isRequired,
+		updateValue: PropTypes.func,
+		error: PropTypes.oneOfType( [ PropTypes.string, PropTypes.bool ] ),
+		className: PropTypes.string,
+		options: PropTypes.array,
+	};
+
+	render() {
+		const { id, title, description, value, placeholder, error, className, options } = this.props;
+
+		return (
+			<FormFieldset className={ className }>
+				<FormLabel htmlFor={ id }>{ title }</FormLabel>
+				<TokenField
+					id={ id }
+					name={ id }
+					placeholder={ placeholder }
+					value={ value.map( this.getSlugFromId.bind( this ) ) }
+					suggestions={ options.map( option => option.slug ) }
+					isError={ Boolean( error ) }
+					onChange={ this.onChange.bind( this ) }
+					displayTransform={ this.transformForDisplay.bind( this ) }
+				/>
+				{ error && typeof error === 'string' && <FieldError text={ error } /> }
+				{ ! error &&
+					description && <FormSettingExplanation>{ description }</FormSettingExplanation> }
+			</FormFieldset>
+		);
+	}
+
+	getSlugFromId( id ) {
+		const found = this.props.options.find( option => {
+			return option.id === id;
+		} );
+
+		return found ? found.slug : null;
+	}
+
+	getIdFromSlugOrName( slug ) {
+		// Try a percise slug match first (if an item was clicked or Enter was pressed)
+		let found = this.props.options.find( option => {
+			return option.slug === slug;
+		} );
+
+		// If there are no precise matches, try guessing
+		if ( ! found && slug.trim().length ) {
+			found = this.props.options.find( option => {
+				return 0 === option.name.toLowerCase().indexOf( slug.trim().toLowerCase() );
+			} );
+		}
+
+		return found ? found.id : null;
+	}
+
+	transformForDisplay( token ) {
+		const option = this.props.options.find( item => item.slug === token );
+
+		return option ? option.name : token;
+	}
+
+	onChange( strings ) {
+		const { updateValue } = this.props;
+
+		const updatedValue = uniqBy(
+			strings.map( this.getIdFromSlugOrName.bind( this ) ).filter( item => null !== item )
```

Comment:
Why do we need to match on anything other than slug here?

Reply:
This was one of the overcomplications caused by using slugs for the token field instead of names: If someone entered or selected the name of a class (ex. the Bulgarian "Книга") it wouldn't have worked if I were to only check slugs and not names.

Now that I am using names as suggestions, only `getIdFromName` is necessary.